keps/2941-DRA-Structured-Parameters/README.md

```diff
@@ -0,0 +1,416 @@
+1. For DRAPrioritizedLists: all the mentioned device classes in the request will be counted against the quota
+2. For DRAAdminAccess: This feature can only be enabled in admin namespace, therefore it should be skipped for being 
+   counted against ClusterQuota. This is a different stance than Kubernetes ResourceQuota because kubernetes ResourceQuota is
+   namespace scoped. As a result, admin users can account for quota independent of user workloads. On the contrary, with 
+   Kueue, since quota is part of ClusterQuota a cluster scoped object, admin workloads using devices in admin namespace
+   if counted against quota, will eat up quota meant for user workloads.
+3. For ResourceClaims with allocation mode `All`: worst-case scenario of the max number of devices that could allocated 
+   to a single claim will be used against quota.
+
+
+## Design Details
+
+### Resource Quota API
+
+```golang
+type ResourceQuota struct {
+    // kind is used to configure if this is for a DRA Device. Its value will be DeviceClass.
+    // +featureGate=KueueDynamicResourceAllocation
+    Kind *string `json:"kind,omitempty"`
+    
+    // deviceClassNames lists the names of all the device classes that will count against
+    // the quota defined in this resource quota object.
+    // +listType=atomic
+    // +optional
+    // +featureGate=KueueDynamicResourceAllocation
+    DeviceClassNames []corev1.ResourceName `json:"deviceClassNames,omitempty"`
+}
+```
+
+Kind field in ResourceQuota allows Kueue to distinguish between a Core resource and a DeviceClass. When the value of kind
+field is configured to be DeviceClass, the `name` becomes a canonical name of the collection of resources that can be 
+provisioned for each device class present in `deviceClassNames` field.
+
+
+With this, a ClusterQueue could be defined as follows:
+
+```yaml
+apiVersion: kueue.x-k8s.io/v1beta1
+kind: ClusterQueue
+metadata:
+  name: "cluster-queue"
+spec:
+  namespaceSelector: {} # match all.
+  resourceGroups:
+  - coveredResources: ["cpu", "memory", "gpu.example.com"]
+    flavors:
+    - name: "default-flavor"
+      resources:
+      - name: "cpu"
+        nominalQuota: 9
+      - name: "memory"
+        nominalQuota: "200Mi"
+      - name: "single-gpu"
+        deviceClassNames: ["gpu.example.com"]
+        nominalQuota: 2
+        kind: "DeviceClass"
+      - name: "shared-gpus"
+        deviceClassNames: 
+        - "ts-shared.gpu.example.com"
+        - "sp-shared.gpu.example.com"
+        nominalQuota: 2
+        kind: "DeviceClass"
+```
+
+The above ClusterQueue is an example configuration of a queue, with half quota configured for single allocation of example
+GPUs, and half quota configured for GPUs that are shared by workloads. Similarly, when DRAPartitionableDevices feature
+is supported in kubernetes, GPUs partitions can be represented by a single device class. This allows for quota to be 
+configured in Kueue without any modifications. 
+
+### Workloads
```

Comment:
Workload object should contain a snapshot of the total amount of resources the job requested at admission time. If, at any point in time, the real resource request changes, the workload object should be re-evaluated, re-created and re-admitted.

Reply:
That will be enough, thanks!

Reply:
sgtm